signature/src/lib.rs

```diff
@@ -30,3 +35,377 @@ pub extern "C" fn verify_rsa(
+
+    #[test]
+    fn test_verify_succeeds_if_valid() {
+        let r = parse_and_verify(
+            VALID_CERT_CHAIN,
+            VALID_INPUT,
+            VALID_SIGNATURE,
+            1615559719, // March 12, 2021
+            ROOT_HASH,
+            VALID_HOSTNAME,
+        );
+        assert!(r.is_ok(), "Found unexpected error: {}", r.unwrap_err());
+    }
+
+    #[test]
+    fn test_verify_fails_if_root_hash_mismatch() {
+        let r = parse_and_verify(
+            VALID_CERT_CHAIN,
+            VALID_INPUT,
+            VALID_SIGNATURE,
+            1615559719, // March 12, 2021
+            &ROOT_HASH.replace("A", "B"),
+            VALID_HOSTNAME,
+        );
+        assert_eq!(r, Err(BalrogError::RootHashMismatch));
+    }
+
+    #[test]
+    fn test_verify_fails_if_cert_invalid() {
+        let r = parse_and_verify(
+            INVALID_CERTIFICATE,
+            VALID_INPUT,
+            VALID_SIGNATURE,
+            1615559719, // March 12, 2021
+            ROOT_HASH,
+            VALID_HOSTNAME,
+        );
+        /* I think this is the wrong error code actually...  */
+        /* I would have expected X509Error::InvalidCertificate */
+        assert_eq!(r, Err(BalrogError::PemDecodeError));
+    }
+
+    #[test]
+    fn test_verify_fails_if_hostname_mismatch() {
+        let r = parse_and_verify(
+            VALID_CERT_CHAIN,
+            VALID_INPUT,
+            VALID_SIGNATURE,
+            1615559719, // March 12, 2021
+            ROOT_HASH,
+            "example.com",
+        );
+        assert_eq!(r, Err(BalrogError::HostnameMismatch));
+    }
+
+    #[test]
+    fn test_verify_fails_if_cert_has_expired() {
+        let r = parse_and_verify(
+            VALID_CERT_CHAIN,
+            VALID_INPUT,
+            VALID_SIGNATURE,
+            1215559719, // July 9, 2008
+            ROOT_HASH,
+            VALID_HOSTNAME,
+        );
+        assert_eq!(r, Err(BalrogError::CertificateExpired));
+    }
+
+    #[test]
+    fn test_verify_fails_if_bad_siganture() {
+        let r = parse_and_verify(
+            VALID_CERT_CHAIN,
+            VALID_INPUT,
+            &VALID_SIGNATURE.to_ascii_lowercase(),
+            1615559719, // March 12, 2021
+            ROOT_HASH,
+            VALID_HOSTNAME,
+        );
+        assert_eq!(r, Err(BalrogError::from(X509Error::SignatureVerificationError)));
+    }
+
+    #[test]
+    fn test_verify_prod_example() {
+        let r = parse_and_verify(
+            PROD_CERT_CHAIN,
+            PROD_INPUT_DATA,
+            PROD_SIGNATURE,
+            1696272799, // Oct 2, 2023 - the time when I wrote this test
+            PROD_ROOT_HASH,
+            PROD_HOSTNAME
+        );
+        assert!(r.is_ok(), "Found unexpected error: {}", r.unwrap_err());
```

Comment:
Could a comment be added explaining why this is test ie expected to fail?

Reply:
This test is expecting to pass, the remaining arguments to `assert!()` are what to display upon a failed assert.

Reply:
Ah, missed the !. Thanks.